lib/ansible/modules/network/fortimanager/fmgr_fwobj_service.py

```diff
@@ -0,0 +1,752 @@
+
+    datagram = fmgr_del_none(datagram)
+
+    if paramgram["mode"] == "set":
+        response = fmg.set(url, datagram)
+        # IF MODE = ADD  -- USE THE 'ADD' API CALL MODE
+    if paramgram["mode"] == "add":
+        response = fmg.add(url, datagram)
+        # IF MODE = DELETE  -- USE THE DELETE URL AND API CALL MODE
+    if paramgram["mode"] == "delete":
+        response = fmg.delete(url, datagram)
+
+    return response
+
+
+def fmgr_fwobj_service_group(fmg, paramgram):
+    """
+    # NOTES
+    only advanced option is color
+    when explicit proxy is set no other options are presented
+    add members list and boom
+    explicit-proxy = 0 is default
+    meta fields = {}
+    color =
+    comment
+    """
+    response = (-100000, {"msg": "Nothing Happened."})
+    if paramgram["mode"] in ['set', 'add']:
+        url = '/pm/config/adom/{adom}/obj/firewall/service/group'.format(adom=paramgram["adom"])
+        datagram = {
+            "name": paramgram["group-name"],
+            "comment": paramgram["comment"],
+            "proxy": paramgram["explicit-proxy"],
+            "color": paramgram["color"]
+        }
+
+        members = paramgram["group-member"]
+        member = []
+        for obj in members.split(","):
+            member.append(obj.strip())
+        datagram["member"] = member
+
+    if paramgram["mode"] == "delete":
+        datagram = {
+            "name": paramgram["name"]
+        }
+        # SET DELETE URL
+        url = '/pm/config/adom/{adom}/obj/firewall/service/group' \
+              '/{name}'.format(adom=paramgram["adom"], name=paramgram["group-name"])
+
+    datagram = fmgr_del_none(datagram)
+
+    if paramgram["mode"] == "set":
+        response = fmg.set(url, datagram)
+        # IF MODE = ADD  -- USE THE 'ADD' API CALL MODE
+    if paramgram["mode"] == "add":
+        response = fmg.add(url, datagram)
+        # IF MODE = DELETE  -- USE THE DELETE URL AND API CALL MODE
+    if paramgram["mode"] == "delete":
+        response = fmg.delete(url, datagram)
+
+    return response
+
+
+def fmgr_fwobj_service_category(fmg, paramgram):
+    """
+    # NOTES
+    """
+    response = (-100000, {"msg": "Nothing Happened."})
+    if paramgram["mode"] in ['set', 'add']:
+        url = '/pm/config/adom/{adom}/obj/firewall/service/category'.format(adom=paramgram["adom"])
+        # GET RID OF ANY WHITESPACE
+        category = paramgram["category"]
+        category = category.strip()
+
+        datagram = {
+            "name": paramgram["category"],
+            "comment": "Created by Ansible"
+        }
+
+    # IF MODE = DELETE
+    if paramgram["mode"] == "delete":
+        datagram = {
+            "name": paramgram["name"]
+        }
+        # SET DELETE URL
+        url = '/pm/config/adom/{adom}/obj/firewall/service/category' \
+              '/{name}'.format(adom=paramgram["adom"], name=paramgram["category"])
+
+    datagram = fmgr_del_none(datagram)
```

Comment:
This code is repeated many time. Can we not write a function which takes a datagram/url/mode and returns the response ?

Reply:
Because not all modes apply to every function, and it's only repeated on this particular module. We're only talking about 12 lines, here. 

We will take your suggestion into account for future modules, thank you Deepak.